### PR TITLE
Add `pg_table_is_visible` scalar

### PR DESCRIPTION
**Pull request by jeeminso**
## Summary of the changes / Why this improves CrateDB
Resolves https://github.com/crate/crate/issues/7988

## Checklist

 - [x] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: server/src/main/java/io/crate/expression/scalar/postgres/PgTableIsVisibleFunction.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        Roles roles = nodeContext.roles();
+        Role user = roles.findUser(sessionSettings.userName());
+        List<String> searchPathsWithDQLPrivileges = new ArrayList<>();
+        for (String searchPath : sessionSettings.searchPath()) {
+            if (HasSchemaPrivilegeFunction.checkBySchemaOid(
+                roles,
+                user,
+                OidHash.schemaOid(searchPath),
+                List.of(Permission.DQL),
+                schemas)) {
```
Comment (jeeminso) [inline server/src/main/java/io/crate/expression/scalar/postgres/PgTableIsVisibleFunction.java:96]: Looks like Postgres' privilege is a bit different:
```
-- 'usage' on schema does not imply the tables within 
=> select * from s2.t2;
ERROR:  permission denied for table t2
-- but still can perform `pg_table_is_visible`
=> select pg_table_is_visible('s2.t2'::regclass::oid);
 pg_table_is_visible 
---------------------
 t
(1 row)

```
And checking for dql on schema aligned with Postgres' behaviour but not consistent with CrateDB. I think we should check for table level privilege instead for CrateDB.

File: server/src/main/java/io/crate/expression/scalar/postgres/PgTableIsVisibleFunction.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        if (relationName == null) {
+            return false;
+        }
+
+        SessionSettings sessionSettings = txnCtx.sessionSettings();
+        Roles roles = nodeContext.roles();
+        Role user = roles.findUser(sessionSettings.userName());
+        for (String searchPathIdx : sessionSettings.searchPath()) {
+            try {
+                schemas.findRelation(
```
Comment (jeeminso) [inline server/src/main/java/io/crate/expression/scalar/postgres/PgTableIsVisibleFunction.java:86]: Handling foreign tables, views, etc is done by `findRelation`

File: server/src/main/java/io/crate/expression/scalar/postgres/PgTableIsVisibleFunction.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+                    QualifiedName.of(searchPathIdx, relationName.name()),
+                    Operation.READ,
+                    user,
+                    SearchPath.createSearchPathFrom(searchPathIdx));
+            } catch (RelationUnknown e) {
+                continue;
+            }
```
Comment (jeeminso) [inline server/src/main/java/io/crate/expression/scalar/postgres/PgTableIsVisibleFunction.java:93]: Depending on try-catch mechanism is not suggested but thought it is too much to come up with a new method in `schemas` just for a scalar function.

**Comment by jeeminso** [comment]
retest this please

File: docs/general/builtins/scalar-functions.rst [hunk 4]
```diff
@@ -4278,6 +4278,31 @@ Example:
     SELECT 1 row in set (... sec)
 
 
+.. _scalar-pg_table_is_visible:
+
+``pg_table_is_visible()``
+-------------------------
+
+The function ``pg_table_is_visible`` returns ``true`` if the current user holds
+``DQL`` privilege on the table and there are no other tables with the same
```
Comment (matriv) [inline docs/general/builtins/scalar-functions.rst:4287]: Have you double-checked with postgres?
What happens if the user has `DML` or `DDL` and not `DQL`, does it return false?

File: docs/general/builtins/scalar-functions.rst [hunk 5]
```diff
@@ -4278,6 +4278,31 @@ Example:
...
+table name(also with ``DQL``) but with different schema names appearing earlier
```
Comment (matriv) [inline docs/general/builtins/scalar-functions.rst:4288]: Here you refer to `name` but we should clarify that the argument is only `oid`, you cannot use the name of the table/view as string.

File: docs/general/builtins/scalar-functions.rst (hunk 4 above)
Comment (jeeminso) [reply docs/general/builtins/scalar-functions.rst:4287]: > Have you double-checked with postgres?

Privileges seem to work differently for Postgres. `select` on table without `usage` on schema, one cannot select on the table(but CrateDB only enforces table level privilege to query tables). Similarly, for `pg_table_is_visible`, Postgres requires `usage` on schema but I thought for CrateDB we should just require table level privilege (DQL). We could also have a team discussion.

> What happens if the user has DML or DDL and not DQL, does it return false?

Yes, but now I see that it should return `true`. If one can execute any query on a table with unqualified name, it is equivalent to 'visible'.

File: server/src/test/java/io/crate/expression/scalar/postgres/PgTableIsVisibleFunctionTest.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+                Set.of(
+                    new Privilege(Policy.GRANT, Permission.DQL, Securable.TABLE, "my_schema2.my_table", "crate")
+                ),
+                null
+            ),
+            List.of(),
+            schemas,
+            searchPaths);
+
+        // my_schema1.my_table is NOT visible
```
Comment (matriv) [inline server/src/test/java/io/crate/expression/scalar/postgres/PgTableIsVisibleFunctionTest.java:179]:
```suggestion
        // my_schema1.my_table is NOT visible, use is missing privileges
```

**Review by matriv — APPROVED** [review]
thx @jeeminso !

Left some minor wording comments.

File: docs/general/builtins/scalar-functions.rst [hunk 7]
```diff
@@ -4297,6 +4297,32 @@ Example:
...
+the same table name and privileges but with different schema names appearing
+earlier in the search path.
+
+Returns: ``boolean``
+
+Example:
+
+::
+
+    cr> select pg_table_is_visible(912037690) as pg_proc_is_visible;
```
[the remark this answers is not in the document]
Comment (jeeminso) [reply docs/general/builtins/scalar-functions.rst:4317]: Oh I just randomly chose to generate the oid `912037690`.

[the remark this answers is not in the document]
Comment (matriv) [reply docs/general/builtins/scalar-functions.rst:4317]: just name it: `is_visible`, to avoid confusion.

**Pull request merged**